public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/CompletionDataProvider.ts

```diff
@@ -16,7 +16,10 @@ export class CompletionDataProvider {
   constructor(
     private languageProvider: LanguageProvider,
     private historyRef: HistoryRef = { current: [] }
-  ) {}
+  ) {
+    this.queryToLabelKeysCache = new Map();
```

Comment:
Just curious about why did you decide to use map and handle deleting/adding, rather than using `LRUCache` that we use in `LokiLanguageProvider` that handles additional logic for us, and we just need to `set` and `get`?

Reply:
We do already use the LRU cache, but in https://github.com/grafana/grafana/pull/75760 I'm proposing to replace it with the browser cache, so I didn't want to go all-in on LRU in case we want to get away from it.

If we're ok using a bit more memory we could certainly use LRU and/or increase the size of the cache, for this MVP I was going for the fastest/lightest possible implementation.

Swapping out for LRU should be trivial as it has the same interface as Map, we'll just need to delete the bit where we manually delete the oldest element and swap out the Map initializer for LRUCache

Reply:
I guess the question is how many items do we want to cache, and how long do we want to cache them, and how bad are stale records being served throughout the session?

I guess the question becomes: is the goal of this cache to prevent duplicate requests, or to cache every possible request in a session?
If we want to cache "everything" or as much as possible, we should use LRU, or beef up the size of the map.
If we're worried about stale results we should range snap and check the date ranges when checking the cache (although in the `getParserAndLabelKeys` we don't have the time range in scope), or stick to a small cache size so results aren't likely to stick around for a long time.

LRU is ~49K, which IMO is a tad overkill for a map interface that keeps track of how many times each element is hit.